autosklearn/data/feature_validator.py

```diff
@@ -0,0 +1,399 @@
+            # Some checks if feat_type is provided
+            if len(feat_type) != np.shape(X_train)[1]:
+                raise ValueError('Array feat_type does not have same number of '
+                                 'variables as X has features. %d vs %d.' %
+                                 (len(feat_type), np.shape(X_train)[1]))
+            if not all([isinstance(f, str) for f in feat_type]):
+                raise ValueError('Array feat_type must only contain strings.')
+
+            for ft in feat_type:
+                if ft.lower() not in ['categorical', 'numerical']:
+                    raise ValueError('Only `Categorical` and `Numerical` are '
+                                     'valid feature types, you passed `%s`' % ft)
+
+            # Here we register proactively the feature types for
+            # Processing Numpy arrays
+            self.feature_types = feat_type
+
+        self._check_data(X_train)
+
+        if X_test is not None:
+            self._check_data(X_test)
+
+            if np.shape(X_train)[1] != np.shape(X_test)[1]:
+                raise ValueError("The feature dimensionality of the train and test "
+                                 "data does not match train({}) != test({})".format(
+                                     np.shape(X_train)[1],
+                                     np.shape(X_test)[1]
+                                 ))
+
+    def _check_data(
+        self,
+        X: typing.Union[np.ndarray, pd.DataFrame],
+    ) -> None:
+        """
+        Feature dimensionality and data type checks
+
+        Parameters
+        ----------
+            X: typing.Union[np.ndarray, pd.DataFrame]
+                A set of features that are going to be validated (type and dimensionality
+                checks) and a encoder fitted in the case the data needs encoding
+        """
+
+        if not isinstance(X, (np.ndarray, pd.DataFrame, list)) and not scipy.sparse.issparse(X):
+            raise ValueError("Auto-sklearn only supports Numpy arrays, Pandas DataFrames,"
+                             " scipy sparse and Python Lists, yet, the provided input is"
+                             " of type {}".format(
+                                 type(X)
+                             ))
+
+        if isinstance(X, list):
+            try:
+                np.array(X).astype(np.float64)
+            except ValueError as e:
+                raise ValueError("When providing a list of features to Auto-sklearn, "
+                                 "this list of elements must contain only numerical values "
+                                 "as it will be converted to Numpy. Casting to Numpy failed "
+                                 "with exception: {}".format(e))
+
+        if self.data_type is None:
+            self.data_type = type(X)
+        if self.data_type != type(X):
+            warnings.warn("Auto-sklearn previously received features of type %s "
```

Comment:
Could you please pass a logger and use it for the warning here?

Reply:
Done!